File: src/main/java/net/Chidoziealways/everythingjapanese/jutsu/JutsuCapability.java
package net.Chidoziealways.everythingjapanese.jutsu;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.NbtOps;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;

public class JutsuCapability implements IJutsuCapability{
    private static final Logger log = LoggerFactory.getLogger(JutsuCapability.class);
    private Set<String> learnedJutsus;
    private String selectedJutsu;

    public JutsuCapability() {
        learnedJutsus = new HashSet<>();
        selectedJutsu = null;
    }

    public JutsuCapability(Set<String> learnedJutsus, String selectedJutsu) {
        this.learnedJutsus = learnedJutsus != null ? learnedJutsus : new HashSet<>();
        this.selectedJutsu = selectedJutsu != null && Objects.requireNonNull(learnedJutsus).contains(selectedJutsu) ? selectedJutsu :
                (!this.learnedJutsus.isEmpty() ? this.learnedJutsus.iterator().next() : " ");
        log.debug("Capability Instance During Deserialization: {}", this);
    }
    
    @Override
    public void learnJutsu(String jutsuId) {
        learnedJutsus.add(jutsuId);
        log.info("learnt jutsu: {}", jutsuId);
        selectedJutsu = jutsuId;
    }

    public void learnJutsu(Jutsu jutsu) {
        learnedJutsus.add(jutsu.getId().getPath());
    }

    @Override
    public boolean hasLearnedJutsu(String jutsuId) {
        return learnedJutsus.contains(jutsuId);
    }

    @Override
    public Set<String> getLearnedJutsus() {
        return learnedJutsus;
    }

    @Override
    public String getSelectedJutsu() {
        return selectedJutsu;
    }

    @Override
    public void cycleJutsu() {
        log.info("Hello from Before Return");
        log.debug("Capability Instance During Cycle Jutsu: {}", this);

        if (learnedJutsus == null) {
            log.error("learned jutsus is NULL!");
            return;
        }

        if (learnedJutsus.isEmpty()) {
            log.warn("No Learnt Jutsus, returning! {}", learnedJutsus);
            return;
        }

        log.info("Hello from cycle");

        List<String> jutsuList = new ArrayList<>(learnedJutsus);

        if (selectedJutsu == null || !getLearnedJutsus().contains(selectedJutsu)) {
            selectedJutsu = jutsuList.getFirst(); // Default to first Jutsu if null
            log.info("selectedJutsu was null, setting to first Jutsu: {}", selectedJutsu);
            return;
        }

        int index = jutsuList.indexOf(selectedJutsu);
        index = (index + 1) % jutsuList.size();
        selectedJutsu = jutsuList.get(index);
        log.info("Cycled to Jutsu: {}", selectedJutsu);
    }

    @Override
    public CompoundTag serializeNBT() {
        return (CompoundTag) JutsuCapability.CODEC.encodeStart(NbtOps.INSTANCE, this)
                .resultOrPartial(error -> System.err.println("Failed to Serialize JutsuCapability: " + error))
                .orElse(new CompoundTag());
    }

    @Override
    public void deserializeNBT(CompoundTag nbt) {
        log.info("Deserializing Jutsu NBT: {}", nbt);

        JutsuCapability.CODEC.decode(NbtOps.INSTANCE, nbt)
                .resultOrPartial(error -> System.err.println("Failed to Deserialize JutsuCapability: " + error))
                .ifPresent(iJutsuCapabilityTagPair -> {
                    IJutsuCapability capability = iJutsuCapabilityTagPair.getFirst();

                    log.info("Decoded JutsuCapability: {}", capability.getLearnedJutsus());

                    this.learnedJutsus = new HashSet<>(capability.getLearnedJutsus());
                    this.selectedJutsu = nbt.contains("selectedJutsu") ? nbt.getStringOr("selectedJutsu", " ") :
                            (!this.learnedJutsus.isEmpty() ? this.learnedJutsus.iterator().next() : " ");

                    log.info("Final Learned Jutsus after Deserialization: {}", this.learnedJutsus);
                });
    }

    public static final Codec<IJutsuCapability> CODEC = RecordCodecBuilder.create(instance -> instance.group(
            Codec.STRING.listOf().fieldOf("learnedJutsus").forGetter(o -> o.getLearnedJutsus().stream().toList()),
            Codec.STRING.fieldOf("selectedJutsu").forGetter(o -> o.getSelectedJutsu() != null ? o.getSelectedJutsu() : " ")
    ).apply(instance, (learnedJutsus, selectedJutsu) -> new JutsuCapability(Set.copyOf(learnedJutsus), selectedJutsu)));
}

File: src/main/kotlin/net/Chidoziealways/everythingjapanese/jutsu/JutsuCapability.kt
package net.Chidoziealways.everythingjapanese.jutsu

import com.mojang.datafixers.util.Pair
import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.NbtOps
import net.minecraft.nbt.Tag
import net.minecraft.server.level.ServerPlayer
import net.neoforged.neoforge.network.PacketDistributor
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*
import java.util.Set
import java.util.function.Consumer
import java.util.function.Function

class JutsuCapability : IJutsuCapability {
    private var learnedJutsus: MutableSet<String> = HashSet()
    private var selectedJutsu: String = ""
    private var jutsuMastery: MutableMap<String, Float> = HashMap()

    constructor() {
        learnedJutsus = HashSet()
        selectedJutsu = ""
        jutsuMastery = HashMap()
    }

    constructor(learnedJutsus: MutableSet<String>, selectedJutsu: String?, masteryMap: MutableMap<String, Float> = HashMap()) {
        this.learnedJutsus = learnedJutsus
        this.jutsuMastery = masteryMap
        this.selectedJutsu = if (selectedJutsu != null && Objects.requireNonNull<MutableSet<String>>(learnedJutsus)
                .contains(selectedJutsu)
        ) selectedJutsu else (if (!this.learnedJutsus.isEmpty()) this.learnedJutsus.iterator().next() else " ")
        log.debug("Capability Instance During Deserialization: {}", this)
    }

    override fun learnJutsu(jutsuId: String, player: ServerPlayer) {
        learnedJutsus.add(jutsuId)
        jutsuMastery.putIfAbsent(jutsuId, 0f)
        selectedJutsu = jutsuId
        PacketDistributor.sendToPlayer(player, JutsuSyncPacket(learnedJutsus, jutsuMastery, selectedJutsu))
        log.info("learnt jutsu: {}", jutsuId)
    }

    override fun hasLearnedJutsu(jutsuId: String): Boolean {
        return learnedJutsus.contains(jutsuId)
    }

    override fun getLearnedJutsus(): MutableSet<String> {
        return learnedJutsus
    }

    override fun setLearnedJutsus(jutsus: MutableSet<String>) {
        learnedJutsus = jutsus
    }

    override fun getMastery(jutsuId: String): Float {
        return jutsuMastery[jutsuId] ?: 0f
    }

    override fun addMastery(jutsuId: String, amount: Float, player: ServerPlayer) {
        val current = jutsuMastery.getOrDefault(jutsuId, 0f)
        val updated = (current + amount).coerceAtMost(1000f)
        jutsuMastery[jutsuId] = updated
        PacketDistributor.sendToPlayer(player, JutsuSyncPacket(learnedJutsus, jutsuMastery, selectedJutsu))
    }

    override fun getMastery(): Map<String, Float> {
        return jutsuMastery
    }

    override fun setMastery(masteryMap: MutableMap<String, Float>) {
        jutsuMastery = masteryMap
    }

    override fun getSelectedJutsu(): String {
        return selectedJutsu
    }

    override fun setSelectedJutsu(jutsu: String) {
        selectedJutsu = jutsu
    }

    override fun cycleJutsu(player: ServerPlayer) {
        log.info("Hello from Before Return")
        log.debug("Capability Instance During Cycle Jutsu: {}", this)

        if (learnedJutsus.isEmpty()) {
            log.warn("No Learnt Jutsus, returning! {}", learnedJutsus)
            return
        }

        log.info("Hello from cycle")

        val jutsuList: MutableList<String> = ArrayList<String>(learnedJutsus)

        if (!getLearnedJutsus().contains(selectedJutsu)) {
            selectedJutsu = jutsuList.first() // Default to first Jutsu if null
            log.info("selectedJutsu was null, setting to first Jutsu: {}", selectedJutsu)
            return
        }

        var index = jutsuList.indexOf(selectedJutsu)
        index = (index + 1) % jutsuList.size
        selectedJutsu = jutsuList[index]
        log.info("Cycled to Jutsu: {}", selectedJutsu)
        PacketDistributor.sendToPlayer(player, JutsuSyncPacket(learnedJutsus, jutsuMastery, selectedJutsu))
    }

    override fun serializeNBT(): CompoundTag {
        return CODEC.encodeStart<Tag>(NbtOps.INSTANCE, this)
            .resultOrPartial { error: String -> System.err.println("Failed to Serialize JutsuCapability: $error") }
            .orElse(CompoundTag()) as CompoundTag
    }

    override fun deserializeNBT(nbt: CompoundTag) {
        log.info("Deserializing Jutsu NBT: {}", nbt)

        CODEC.decode(NbtOps.INSTANCE, nbt)
            .resultOrPartial { error -> log.error("Failed to deserialize JutsuCapability: {}", error) }
            .ifPresent { pair ->
                val cap = pair.first ?: run {
                    log.error("Decoded capability is null! Keeping old data.")
                    return@ifPresent
                }
                this.learnedJutsus = HashSet(cap.getLearnedJutsus())
                this.jutsuMastery = HashMap(cap.getMastery())
                this.selectedJutsu = if (cap.getSelectedJutsu().isNotBlank() && learnedJutsus.contains(cap.getSelectedJutsu()))
                    cap.getSelectedJutsu() else learnedJutsus.firstOrNull() ?: ""
            }

    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(JutsuCapability::class.java)
        val CODEC: Codec<IJutsuCapability> =
            RecordCodecBuilder.create<IJutsuCapability>(Function { instance: RecordCodecBuilder.Instance<IJutsuCapability> ->
                instance.group(
                    Codec.STRING.listOf().fieldOf("learnedJutsus").forGetter<IJutsuCapability> { o: IJutsuCapability ->
                        o.getLearnedJutsus().stream().toList()
                        },
                    Codec.STRING.fieldOf("selectedJutsu").forGetter<IJutsuCapability> { o: IJutsuCapability -> o.getSelectedJutsu() },
                    Codec.unboundedMap(Codec.STRING, Codec.FLOAT).fieldOf("masteryMap")
                        .forGetter { it.getMastery() }
                ).apply<IJutsuCapability>(
                    instance
                ) { learnedJutsus: MutableList<String>, selectedJutsu: String, mastery: Map<String, Float> ->
                    JutsuCapability(
                        Set.copyOf<String>(learnedJutsus), selectedJutsu, mastery.toMutableMap()
                    )
                }
            })
    }
}
